sqlmesh/core/model/definition.py

```diff
@@ -1492,6 +1492,14 @@ def load_sql_based_model(
         meta_fields["dialect"] = meta_fields["dialect"].name
 
     name = meta_fields.pop("name", "")
+
+    # If the sql file is under a schema directory infer the name from path
+    if not name:
+        name = path.stem
+        subdirectories = path.parts[path.parts.index("models") + 1 : -1]
+        if subdirectories:
+            name = ".".join(subdirectories) + "." + name
```

Comment:
What if the nestedness is more than 3 levels deep? Should we fail then? Or only take the last 3 parts?

How about Python models?

Reply:
For python models I overriden the `__call__` method in the model `decorator` to derive the name from the path there, I'm open to suggestions though if we should handle it in a better way.

For the nestedness I adapted the code to take the last 3 parts.